### ISSUE TITLE
CMake MinGW compilation issues

### ISSUE THREAD

**Issue by Zapeth** [Arisotura/melonDS#307]
Had to make some changes to CMakeLists.txt in order to be able to compile with MinGW (MSYS):

- (Re)add `src/libui_sdl/libui/windows/entry.cpp`, otherwise build would fail with undefined references

- add `LINK_LIBRARIES("uxtheme")`, otherwise build would fail with an undefined reference to `SetWindowTheme`

- add include directory for libpcap files (not sure whats the recommended way to add this, I used an environment variable for the include path and define it before executing cmake because I don't have libpcap installed as a system package)

Besides these changes in the CMakeLists.txt I also had to add `#define HAVE_REMOTE` before `#include <pcap/pcap.h>` in `src/libui_sdl/Platform.cpp`, or some error would show up during compilation.

**Comment by StapleButter** [comment]
can ya submit the CMakeLists change as a pull request, would be cool :)

as for the pcap shit I need to look into it anyway

**Comment by Zapeth** [comment]
Sure, I was just wondering if the first point was intentional or not (since its already in the file but commented out), especially since there are Windows binaries of this version available on your site.

Meaning you managed to compile it in its current state and I'm just doing it wrong/differently, or you have some local changes regarding this that haven't been pushed to master branch (in the latter case it might be better if I don't do a PR and you push your outstanding changes at some point).

> as for the pcap shit I need to look into it anyway
In that case I'll skip the pcap related stuff, I'm not really sure if its required for other setups than mine anyway (at least regarding the `#define HAVE_REMOTE`, but even for the libpcap include path there's probably a better solution).

**Comment by Ace4896** [comment]
Now that #398 is merged, do you still have these compilation issues? Just asking since it revamped the CMake system and fixed some things related to MSYS.

**Comment by Zapeth** [comment]
Unfortunately yes, it seems that linking the static libraries failes:
```
...
[ 21%] Building CXX object src/CMakeFiles/core.dir/Wifi.cpp.obj
[ 23%] Building CXX object src/CMakeFiles/core.dir/WifiAP.cpp.obj
[ 23%] Linking CXX static library libcore.a
C:/msys64/mingw64/bin/ranlib.exe: unable to rename 'libcore.a'; reason: File exists
make[2]: *** [src/CMakeFiles/core.dir/build.make:385: src/libcore.a] Error 1
make[2]: *** Deleting file 'src/libcore.a'
make[1]: *** [CMakeFiles/Makefile2:91: src/CMakeFiles/core.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```
I searched a bit around online myself, but the most common solutions for this issue are not applicable for me (working in paths with spaces or some anti-virus program interefering).

However I have a suspicion that it might be due to outdated packages on my side (gcc 8.2.0, cmake 3.12.1), what are your versions of these packages?

Unfortunately I can't simply try updating my packages right now because it would break other stuff for me, but if a third person can verify that it works for them with an up-to-date msys2 setup then I can close this for the time being (its not like I was waiting for this to be resolved anyway) and report back once I get around to updating my msys2 setup.

**Comment by Ace4896** [comment]
I used GCC 8.3.0 and CMake 3.13.4 to test the PR I mentioned; I just tried with the latest CMake (3.14.2) and it compiles just fine for me. Maybe the versions aren't the issue? (in Ubuntu, I'm still using GCC 7.x and CMake 3.10.2)

The only other thing I can think of is file permissions (e.g. set to read-only). Just to make sure, was the output you pasted from a newly generated makefile?

**Comment by Zapeth** [comment]
Yes, the repository was also freshly cloned. After it failed the first time I also tried it with the same commands from the readme and with a cmake Debug configuration, but same error occurred every time.

My suspicion in regards to the outdated packages being the issue is more in terms of regression issues in those versions than missing features or existing bugs (since I seem to recall reading about a couple of issues with GCC when the upgrade to 8 happened on MinGW), but it could very well be some other issue too.

**Comment by kwkam** [comment]
I am using this CMakeLists to cross-compile, please see if it helps you:
```cmake
cmake_minimum_required(VERSION 2.8.11)
project(melonDS)

set(CMAKE_CXX_STANDARD 11)

if(NOT CMAKE_BUILD_TYPE)
    set(CMAKE_BUILD_TYPE Release)
endif()

if(CMAKE_BUILD_TYPE STREQUAL Debug)
    set(CMAKE_CXX_FLAGS "-g")
elseif(CMAKE_BUILD_TYPE STREQUAL Release)
    set(CMAKE_CXX_FLAGS "-O3 -flto")
endif()

find_package(PkgConfig REQUIRED)

set(SRC_MAIN
    src/libui_sdl/main.cpp
    src/libui_sdl/Platform.cpp
    src/libui_sdl/PlatformConfig.cpp
    src/libui_sdl/LAN_Socket.cpp
    src/libui_sdl/LAN_PCap.cpp
    src/libui_sdl/DlgAudioSettings.cpp
    src/libui_sdl/DlgEmuSettings.cpp
    src/libui_sdl/DlgInputConfig.cpp
    src/libui_sdl/DlgWifiSettings.cpp
    src/ARM.cpp
    src/ARMInterpreter.cpp
    src/ARMInterpreter_ALU.cpp
    src/ARMInterpreter_Branch.cpp
    src/ARMInterpreter_LoadStore.cpp
    src/Config.cpp
    src/CP15.cpp
    src/CRC32.cpp
    src/DMA.cpp
    src/GPU.cpp
    src/GPU2D.cpp
    src/GPU3D.cpp
    src/GPU3D_Soft.cpp
    src/NDS.cpp
    src/NDSCart.cpp
    src/RTC.cpp
    src/Savestate.cpp
    src/SPI.cpp
    src/SPU.cpp
    src/Wifi.cpp
    src/WifiAP.cpp
)

set(SRC_LIBUI
    src/libui_sdl/libui/common/areaevents.c
    src/libui_sdl/libui/common/control.c
    src/libui_sdl/libui/common/debug.c
    src/libui_sdl/libui/common/matrix.c
    src/libui_sdl/libui/common/shouldquit.c
    src/libui_sdl/libui/common/userbugs.c
)

set(SRC_LIBUI_NIX
    src/libui_sdl/libui/unix/alloc.c
    src/libui_sdl/libui/unix/area.c
    src/libui_sdl/libui/unix/box.c
    src/libui_sdl/libui/unix/button.c
    src/libui_sdl/libui/unix/cellrendererbutton.c
    src/libui_sdl/libui/unix/checkbox.c
    src/libui_sdl/libui/unix/child.c
    src/libui_sdl/libui/unix/colorbutton.c
    src/libui_sdl/libui/unix/combobox.c
    src/libui_sdl/libui/unix/control.c
    src/libui_sdl/libui/unix/datetimepicker.c
    src/libui_sdl/libui/unix/debug.c
    src/libui_sdl/libui/unix/draw.c
    src/libui_sdl/libui/unix/drawmatrix.c
    src/libui_sdl/libui/unix/drawpath.c
    src/libui_sdl/libui/unix/drawtext.c
    src/libui_sdl/libui/unix/editablecombo.c
    src/libui_sdl/libui/unix/entry.c
    src/libui_sdl/libui/unix/fontbutton.c
    src/libui_sdl/libui/unix/form.c
    src/libui_sdl/libui/unix/future.c
    src/libui_sdl/libui/unix/graphemes.c
    src/libui_sdl/libui/unix/grid.c
    src/libui_sdl/libui/unix/group.c
    src/libui_sdl/libui/unix/image.c
    src/libui_sdl/libui/unix/label.c
    src/libui_sdl/libui/unix/main.c
    src/libui_sdl/libui/unix/menu.c
    src/libui_sdl/libui/unix/multilineentry.c
    src/libui_sdl/libui/unix/progressbar.c
    src/libui_sdl/libui/unix/radiobuttons.c
    src/libui_sdl/libui/unix/separator.c
    src/libui_sdl/libui/unix/slider.c
    src/libui_sdl/libui/unix/spinbox.c
    src/libui_sdl/libui/unix/stddialogs.c
    src/libui_sdl/libui/unix/tab.c
    src/libui_sdl/libui/unix/text.c
    src/libui_sdl/libui/unix/util.c
    src/libui_sdl/libui/unix/window.c
)

set(SRC_LIBUI_WIN
    src/libui_sdl/libui/windows/alloc.cpp
    src/libui_sdl/libui/windows/area.cpp
    src/libui_sdl/libui/windows/areadraw.cpp
    src/libui_sdl/libui/windows/areaevents.cpp
    src/libui_sdl/libui/windows/areascroll.cpp
    src/libui_sdl/libui/windows/areautil.cpp
    src/libui_sdl/libui/windows/box.cpp
    src/libui_sdl/libui/windows/button.cpp
    src/libui_sdl/libui/windows/checkbox.cpp
    src/libui_sdl/libui/windows/colorbutton.cpp
    src/libui_sdl/libui/windows/colordialog.cpp
    src/libui_sdl/libui/windows/combobox.cpp
    src/libui_sdl/libui/windows/container.cpp
    src/libui_sdl/libui/windows/control.cpp
    src/libui_sdl/libui/windows/d2dscratch.cpp
    src/libui_sdl/libui/windows/datetimepicker.cpp
    src/libui_sdl/libui/windows/debug.cpp
    src/libui_sdl/libui/windows/draw.cpp
    src/libui_sdl/libui/windows/drawmatrix.cpp
    src/libui_sdl/libui/windows/drawpath.cpp
    src/libui_sdl/libui/windows/drawtext.cpp
    src/libui_sdl/libui/windows/dwrite.cpp
    src/libui_sdl/libui/windows/editablecombo.cpp
    src/libui_sdl/libui/windows/entry.cpp
    src/libui_sdl/libui/windows/events.cpp
    src/libui_sdl/libui/windows/fontbutton.cpp
    src/libui_sdl/libui/windows/fontdialog.cpp
    src/libui_sdl/libui/windows/form.cpp
    src/libui_sdl/libui/windows/graphemes.cpp
    src/libui_sdl/libui/windows/grid.cpp
    src/libui_sdl/libui/windows/group.cpp
    src/libui_sdl/libui/windows/init.cpp
    src/libui_sdl/libui/windows/label.cpp
    src/libui_sdl/libui/windows/main.cpp
    src/libui_sdl/libui/windows/menu.cpp
    src/libui_sdl/libui/windows/multilineentry.cpp
    src/libui_sdl/libui/windows/parent.cpp
    src/libui_sdl/libui/windows/progressbar.cpp
    src/libui_sdl/libui/windows/radiobuttons.cpp
    src/libui_sdl/libui/windows/separator.cpp
    src/libui_sdl/libui/windows/sizing.cpp
    src/libui_sdl/libui/windows/slider.cpp
    src/libui_sdl/libui/windows/spinbox.cpp
    src/libui_sdl/libui/windows/stddialogs.cpp
    src/libui_sdl/libui/windows/tab.cpp
    src/libui_sdl/libui/windows/tabpage.cpp
    src/libui_sdl/libui/windows/text.cpp
    src/libui_sdl/libui/windows/utf16.cpp
    src/libui_sdl/libui/windows/utilwin.cpp
    src/libui_sdl/libui/windows/window.cpp
    src/libui_sdl/libui/windows/winpublic.cpp
    src/libui_sdl/libui/windows/winutil.cpp
    src/libui_sdl/libui/windows/resources.rc
)

if(UNIX)
    list(APPEND SRC_MAIN melon_grc.c)

    add_custom_command(OUTPUT melon_grc.c
        COMMAND glib-compile-resources --sourcedir="${CMAKE_CURRENT_SOURCE_DIR}"
                --target="${CMAKE_CURRENT_BINARY_DIR}/melon_grc.c"
                --generate-source "${CMAKE_CURRENT_SOURCE_DIR}/melon_grc.xml"
        COMMAND glib-compile-resources --sourcedir="${CMAKE_CURRENT_SOURCE_DIR}"
                --target="${CMAKE_CURRENT_BINARY_DIR}/melon_grc.h"
                --generate-header "${CMAKE_CURRENT_SOURCE_DIR}/melon_grc.xml")

    add_library(LIBUI ${SRC_LIBUI} ${SRC_LIBUI_NIX})
    if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
        target_link_libraries(LIBUI "dl")
    endif()

    pkg_check_modules(GTK3 REQUIRED gtk+-3.0)
    target_include_directories(LIBUI PUBLIC ${GTK3_INCLUDE_DIRS})
    target_link_libraries(LIBUI ${GTK3_LIBRARIES})
elseif(WIN32)
    list(APPEND SRC_MAIN melon.rc)

    add_library(LIBUI OBJECT ${SRC_LIBUI} ${SRC_LIBUI_WIN})
    target_link_libraries(LIBUI "comctl32" "d2d1" "dwrite" "usp10" "uxtheme")
    target_link_libraries(LIBUI "iphlpapi" "ws2_32")

    pkg_check_modules(WPCAP REQUIRED wpcap)
    target_include_directories(LIBUI PUBLIC ${WPCAP_INCLUDE_DIRS})
    target_link_libraries(LIBUI ${WPCAP_LIBRARIES})
endif()

pkg_check_modules(SDL2 REQUIRED sdl2)
target_include_directories(LIBUI PUBLIC ${SDL2_INCLUDE_DIR})
target_link_libraries(LIBUI ${SDL2_LIBRARIES})

add_executable(${PROJECT_NAME} ${SRC_MAIN})
target_link_libraries(${PROJECT_NAME} LIBUI)

install(TARGETS ${PROJECT_NAME} RUNTIME DESTINATION ${CMAKE_INSTALL_PREFIX}/bin)
```

**Comment by Zapeth** [comment]
Yes it works (still had to change the wpcap include part though because I was unable to figure out how to tell `pkg_check_modules` where to find the `wpcap` module), but it doesn't really solve the issue of the upstream build configuration/steps not working.

I did notice though that the current CMakeLists
1) never asked or complained about Winpcap directories (are they not required anymore for compilation?) and
2) tries to link static libraries during the compilation (failing for me with the previously posted error), which doesn't happen with the old method (just compiles all files and does a single link generation at the end for the exe).

**Comment by kwkam** [comment]
Where did you install wpcap to? Can you find the file `lib/pkgconfig/wpcap.pc` in your environment?

**Comment by Zapeth** [comment]
I only ever used the [WinPcap Developer's Files](https://www.winpcap.org/devel.htm) (which doesn't seem to feature a `wpcap.pc` file), because I couldn't find an according msys2/mingw package and the official installation doesn't contain the development files.

How did you install WinPcap on your system?

**Comment by kwkam** [comment]
You could try to create one and put it inside the pkgconfig folder, or put it somewhere and set the environment variable PKG_CONFIG_PATH
Here is an example:
```
# Package Information for pkg-config

prefix=C:/path/to/WinPCAP
exec_prefix=${prefix}
libdir=${prefix}/lib
includedir=${prefix}/include

Name: WPCAP
Description: Windows Packet Capture Library
Version: 4.1.final2
Libs: -L${libdir} -lpacket -lwpcap
Libs.private:
Cflags: -I${includedir}
```

**Comment by Pik-a-boo** [comment]
Someone knows how to compile MelonDS with PGO?
How to add the PGO flag and stuff.

**Comment by Zapeth** [comment]
@Ace4896 I finally got around to updating my entire msys2 setup so I tried to compile the most recent master with cmake again.

I didn't run into the aforementioned issue(s), but had some resource file related issues (unable to process spaces in paths and perform a `copy` command with a `*`) which I was able to solve (see https://github.com/Arisotura/melonDS/pull/509).

I'm not sure whether the msys2 updates fixed my other issues or some source file changes since then, but I guess I can close this once the PR is merged.
